libparsec/crates/client/tests/unit/certificates_ops/validate_message.rs

```diff
@@ -0,0 +1,542 @@
+
+    let certificates = env
+        .template
+        .certificates()
+        .map(|e| e.signed)
+        .collect::<Vec<_>>();
+
+    test_register_send_hook(
+        &env.discriminant_dir,
+        move |req: authenticated_cmds::latest::certificate_get::Req| {
+            p_assert_eq!(req.offset, 0);
+
+            authenticated_cmds::latest::certificate_get::Rep::Ok { certificates }
+        },
+    );
+
+    let message = MessageContent::SharingGranted {
+        author: alice.device_id.clone(),
+        timestamp: now,
+        id: vlob_id,
+        encrypted_on: now,
+        name: "wksp1".parse().unwrap(),
+        encryption_revision: 0,
+        key: alice.local_symkey.clone(),
+    };
+
+    let msg = ops
+        .validate_message(
+            3,
+            0,
+            &alice.device_id,
+            now,
+            &message.dump_sign_and_encrypt_for(&alice.signing_key, &alice.public_key()),
+        )
+        .await
+        .unwrap();
+
+    p_assert_eq!(msg, message);
+
+    ops.stop().await;
+}
+
+#[parsec_test(testbed = "minimal")]
+async fn ok_sharing_reencrypted(env: &TestbedEnv) {
+    let vlob_id = VlobID::from_hex("f0000000-0000-0000-0000-000000000001").unwrap();
+    let now = DateTime::from_ymd_hms_us(2020, 1, 1, 0, 0, 0, 0).unwrap();
+
+    let env = env.customize(|builder| {
+        builder
+            .new_user_realm("alice")
+            .customize(|event| event.realm_id = vlob_id);
+    });
+
+    let alice = env.local_device("alice@dev1");
+
+    let ops = certificates_ops_factory(&env, &alice).await;
+
+    let certificates = env
+        .template
+        .certificates()
+        .map(|e| e.signed)
+        .collect::<Vec<_>>();
+
+    test_register_send_hook(
+        &env.discriminant_dir,
+        move |req: authenticated_cmds::latest::certificate_get::Req| {
+            p_assert_eq!(req.offset, 0);
+
+            authenticated_cmds::latest::certificate_get::Rep::Ok { certificates }
+        },
+    );
+
+    let message = MessageContent::SharingReencrypted {
+        author: alice.device_id.clone(),
+        timestamp: now,
+        id: vlob_id,
+        encrypted_on: now,
+        key: alice.local_symkey.clone(),
+        name: "wksp1".parse().unwrap(),
+        encryption_revision: 0,
+    };
+
+    let msg = ops
+        .validate_message(
+            3,
+            0,
+            &alice.device_id,
+            now,
+            &message.dump_sign_and_encrypt_for(&alice.signing_key, &alice.public_key()),
+        )
+        .await
+        .unwrap();
+
+    p_assert_eq!(msg, message);
+
+    ops.stop().await;
+}
+
+#[parsec_test(testbed = "minimal")]
+async fn offline(env: &TestbedEnv) {
+    let now = DateTime::from_ymd_hms_us(2020, 1, 1, 0, 0, 0, 0).unwrap();
+
+    let alice = env.local_device("alice@dev1");
+
+    let ops = certificates_ops_factory(env, &alice).await;
+
+    let err = ops
+        .validate_message(1, 0, &alice.device_id, now, b"")
```

Comment:
Offline because you require a new certificate index that isn't present locally ?